jupyter_scheduler/task_runner.py

```diff
@@ -0,0 +1,310 @@
+    timezone: Optional[str] = None
+    schedule: Optional[str] = None
+
+
+@dataclass
+class JobDefinitionTask:
+    job_definition_id: str
+    next_run_time: int
+
+    def __lt__(self, other):
+        return self.next_run_time < other.next_run_time
+
+
+class PriorityQueue:
+    """A priority queue using heapq"""
+
+    def __init__(self):
+        self._heap = []
+
+    def peek(self):
+        if self.isempty():
+            raise "Queue is empty"
+
+        return self._heap[0]
+
+    def push(self, task: JobDefinitionTask):
+        heappush(self._heap, task)
+
+    def pop(self):
+        task = heappop(self._heap)
+        return task
+
+    def __len__(self):
+        return len(self._heap)
+
+    def isempty(self):
+        return len(self._heap) < 1
+
+
+class Cache:
+    _session = None
+
+    @property
+    def session(self):
+        if not self._session:
+            self.cache_url = "sqlite://"
+            engine = create_engine(self.cache_url, echo=False)
+            Base.metadata.create_all(engine)
+            self._session = sessionmaker(bind=engine)
+
+        return self._session
+
+    def load(self, models: List[DescribeJobDefinitionCache]):
+        with self.session() as session:
+            for model in models:
+                session.add(JobDefinitionCache(**model.dict()))
+            session.commit()
+
+    def get(self, job_definition_id: str) -> DescribeJobDefinitionCache:
+        with self.session() as session:
+            definition = (
+                session.query(JobDefinitionCache)
+                .filter(JobDefinitionCache.job_definition_id == job_definition_id)
+                .first()
+            )
+        return DescribeJobDefinitionCache.from_orm(definition)
+
+    def put(self, model: DescribeJobDefinitionCache):
+        with self.session() as session:
+            session.add(JobDefinitionCache(**model.dict()))
+            session.commit()
+
+    def update(self, job_definition_id: str, model: UpdateJobDefinitionCache):
+        with self.session() as session:
+            session.query(JobDefinitionCache).filter(
+                JobDefinitionCache.job_definition_id == job_definition_id
+            ).update(model.dict(exclude_none=True))
+            session.commit()
+
+
+class TaskRunner:
+    _db_session = None
+    _cache = None
+    _queue = None
+    _scheduler = None
+
+    def __init__(self, scheduler, run_interval: int) -> None:
+        self.run_interval = run_interval
+        self.scheduler = scheduler
+        self.db_session = scheduler.db_session
+
+    @property
+    def cache(self):
+        if not self._cache:
+            self._cache = Cache()
+
+        return self._cache
+
+    @property
+    def queue(self):
+        if not self._queue:
+            self._queue = PriorityQueue()
+
+        return self._queue
+
+    def compute_next_run_time(self, schedule: str, timezone: Optional[str] = None):
+        return compute_next_run_time(schedule, timezone)
```

Comment:
This pattern is useful because it localizes change to one place if something changes, for example if you replaced the inner function with a different one. It is also useful for mocking the method for testing rather than patching the method from utils.